### ISSUE TITLE
Translate & Align V-Pipe Reads

### ISSUE THREAD

**Issue by gordonkoehn** [cbg-ethz/sr2silo#53]
Currently, we use Nextclade to translate from nucleotides to amino acids.

We also use it to realign the notes and amino acids, even though our nucleotides are already aligned.

This seemed fine for small test data, yet it may be infeasible or a mere waste of resources to realign the reads again. 

So consider rewriting the `translate` functions.

**Comment by gordonkoehn** [comment]
From [rneher](https://github.com/rneher)

> A simple script could translate those already aligned reads. You'd need to figure which ORFs you read falls into, the reading frame, and then use something like Bio.Sequence.translate from the biopython package to translate the sequence.

**Comment by gordonkoehn** [comment]
See here.
- https://github.com/nextstrain/nextclade/discussions/1556

**Comment by gordonkoehn** [comment]
So probably we need a custom tool.

**Comment by gordonkoehn** [comment]
@DrYak, for now, I am ignoring insertions completely. There are too many biological unknowns for me here. I'd cherish your advice here sometime.

**Comment by gordonkoehn** [comment]
Need some input before I can continue.

**Comment by gordonkoehn** [comment]
Take-aways from chat with @LaraFuhrmann

- BIG CAVEAT to uploading the `.bam` reads to SILO is that this discards all of V-Pipe's mutation calling efforts, so the database will contain sequencing artefacts. 
- Yet, anyone will still be able to query, which is useful in urgent situations quickly.
- As a rough estimate, V-Pipe's computational effort is 15 % preprocessing / 25 % nucleotide alignment / 60 % mutational calling. 
- The current plan hence is just to take all `.bam` take single reads to align with nextclade and once down, merge pair reads
- Basically, we only use V-Pipe's preprocessing and use Nextclade alignment / at the end, we again have V-Pipe's post-processing.

So actionable: 
- run nextclade for batches of the reads to still run in a small docker, just for longer.

@DrYak FYI. Let's also chat about this before I do it.

**Comment by gordonkoehn** [comment]
Take-aways from chat with @DrYak:

There exists, indeed, no code for this as of now. This is because most of the time, people will use mutations called on nucleotides and only translate these mutations, not entire alignments. 

There are two options:
- 1) build it from the ground( with pysam and proper logic to handle all coding frames, handle corner cases)
- 2) hack Nextclade and hook into the process after the alignment they do

In both cases, the steps would be to take a `.sam` of the single reads and make a `.sam` with paired reads, i.e. using Micha's tool. 

Once that exists, translate the amino acid alignment with either option.

**Comment by gordonkoehn** [comment]
This is a tool that Niko shared 
- [ ] Check out this tool [VIRULINGN](https://github.com/rega-cev/virulign-tutorial)
it appears it does the alignment itself. So this is not what we want, but it reads like there are some difficulties with the translation. Which just proved my point to use a well-supported tool. Then Nextclade will be their better choice if I hack something.

**Comment by gordonkoehn** [comment]
- [ ] How does V-Pipe align?  Is that better in any way than Nextclade?

Ivan mentioned some Probabilititic work going on in the alignment and theorized Nextclade might do something more simplistic.

**Comment by gordonkoehn** [comment]
Managed to get a clear text aligment i.e. equal length to reference as well as insertions per read. 

Now it is ready to pass these strings over to nextclade.

**Comment by gordonkoehn** [comment]
Progress update from Nextclade side
- nextclade is parralelized, it would be hard hack its cli with limited rust knowledge
-  `stripped` contains the input I can provide from the `sr2silo` side. See [here](https://github.com/gordonkoehn/nextclade/blob/11bb4ec8d2dc6ba1fe375465e3f56c5187d6022e/packages/nextclade/src/run/nextclade_run_one.rs#L123)
- translation and alignment of amino acids !! happens [here](https://github.com/gordonkoehn/nextclade/blob/11bb4ec8d2dc6ba1fe375465e3f56c5187d6022e/packages/nextclade/src/translate/translate_genes.rs#L273C1-L286C5)

Obviously realigment has to happen if there are frameshift, this was not obvious to me.

It appears best to import `nextclade` as a `crate` and run these functions as a mini script ?

**Comment by gordonkoehn** [comment]
UPDATE:

Nextclade was the wrong tool for this. 
In principle I could finish this with Nextclade now. 
But BlastX seems to be the correct way to do it, as that's what it was designed for. 

Am cherry-picking all work from before, and migrating to BlastX. See [here](https://github.com/cbg-ethz/sr2silo/pull/88).
